llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+    }
+  };
+
+  // https://alive2.llvm.org/ce/z/rw7XWv
+  // (X&Y)==C ? X|Y : X^Y -> (X^Y)|C : X^Y  or (X^Y)^ C : X^Y
+  // (X&Y)==C ? X^Y : X|Y -> (X|Y)^C : X|Y  or (X|Y)&~C : X|Y
+  if (match(CmpLHS, m_And(m_Value(X), m_Value(Y)))) {
+    if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
+      // (X&Y)==0 ? X|Y : X^Y -> (X^Y)|0 : X^Y -> X^Y
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, XorOps, None) ||
+          // (X&Y)==C ? X|Y : (X^Y)|C -> (X^Y)|C : (X^Y)|C -> (X^Y)|C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, XorOps, None) ||
+          // (X&Y)==C ? X|Y : (X^Y)^C -> (X^Y)^C : (X^Y)^C -> (X^Y)^C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, XorOps, None))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      // (X&Y)==-1 ? X|Y : FV -> -1 : FV
+      else if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, AllOnes);
+    } else if (match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      // (X&Y)==0 ? X^Y : X|Y -> (X|Y)^0 : X|Y -> X|Y
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, OrOps, None) ||
+          // (X&Y)==C ? X^Y : (X|Y)^ C -> (X|Y)^ C : (X|Y)^ C -> (X|Y)^ C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
+          // (X&Y)==C ? X^Y : (X|Y)&~C -> (X|Y)&~C : (X|Y)&~C -> (X|Y)&~C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      // (X&Y)==-1 ? X^Y : FV -> 0 : FV
+      else if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, Null);
+    }
+  }
+
+  // https://alive2.llvm.org/ce/z/5yaUzR
+  // (X|Y)==C ? X&Y : X^Y -> (X^Y)^C : X^Y  or  ~(X^Y)&C : X^Y
+  // (X|Y)==C ? X^Y : X&Y -> (X&Y)^C : X&Y  or  ~(X&Y)&C : X&Y
+  if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y)))) {
+    if (match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y)))) {
+      // (X|Y)==0 ? X&Y : X^Y -> (X^Y)^0 : X^Y -> X^Y
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, XorOps, None) ||
+          // (X|Y)==C ? X&Y: (X^Y)^C -> (X^Y)^C: (X^Y)^C ->  (X^Y)^C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, XorOps, None) ||
+          // (X|Y)==C ? X&Y:~(X^Y)&C ->~(X^Y)&C:~(X^Y)&C -> ~(X^Y)&C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, XorOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      // (X|Y)==0 ? X^Y : X&Y -> (X&Y)^0 : X&Y -> X&Y
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, AndOps, None) ||
+          // (X|Y)==C ? X^Y : (X&Y)^C ->  (X&Y)^C : (X&Y)^C ->  (X&Y)^C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None) ||
+          // (X|Y)==C ? X^Y :~(X&Y)&C -> ~(X&Y)&C :~(X&Y)&C -> ~(X&Y)&C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, AndOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(CmpRHS, m_Zero()))
```

Comment:
Same idea as the `X & Y == -1`, you can be constant folding `X -> 0` and `Y -> 0` for anything.

Maybe add a TODO to move these cases to `foldBinOpIntoSelect`?

Reply:
how about this?
is this implementation good enough to trying?

```
static Instruction *implyBitwiseCond(SelectInst &SI, Value *CmpLHS,
                                     Value *CmpRHS, Value *Val,
                                     InstCombinerImpl &IC) {
  LLVM_DEBUG(dbgs() << "ENTRY : "; CmpLHS->dump(); CmpRHS->dump());
  Instruction *I = dyn_cast<BinaryOperator>(Val);
  Instruction *RetI = I;
  bool Changed = false;
  for (int Idx = 0, Max = I->getNumOperands(); Idx < Max; Idx++) {
    Value *Op = I->getOperand(Idx);
    LLVM_DEBUG(dbgs() << "OPERAND : "; Op->dump());
    if (CmpLHS == Op) {
      RetI = IC.replaceOperand(*RetI, Idx, CmpRHS);
      Changed = true;
    }
    LLVM_DEBUG(dbgs() << "TTTT : "; I->dump());
  }

  LLVM_DEBUG(dbgs() << "TTTT : "; I->dump(); RetI->dump());
  return Changed ? I : 0;
}
```

```
  // (X&Y)==C ? X^Y : X|Y -> (X|Y)^C : X|Y  or (X|Y)&~C : X|Y
  if (match(CmpLHS, m_And(m_Value(X), m_Value(Y)))) {
    // (X&Y)==-1 ? X|Y : FV -> -1 : FV
    if (match(CmpRHS, m_AllOnes())) {
      if (auto I = implyBitwiseCond(SI, X, AllOnes, TrueVal, IC)) {
        if (auto II = implyBitwiseCond(SI, Y, AllOnes, I, IC))
          return IC.replaceOperand(SI, 1, II);
      }
    }
```

Reply:
here is results : 
```
IC: Visiting:   %cond = select i1 %cmp, i32 %or, i32 %xor
ENTRY : i32 %y
i32 -1
OPERAND : i32 %y
TTTT :   %or = or i32 -1, %x
OPERAND : i32 %x
TTTT :   %or = or i32 -1, %x
TTTT :   %or = or i32 -1, %x
  %or = or i32 -1, %x
ENTRY : i32 %x
i32 -1
OPERAND : i32 -1
TTTT :   %or = or i32 -1, %x
OPERAND : i32 %x
TTTT :   %or = or i32 -1, -1
TTTT :   %or = or i32 -1, -1
  %or = or i32 -1, -1
ADD DEFERRED:   %or = or i32 -1, -1
ADD DEFERRED:   %cond = select i1 %cmp, i32 %or, i32 %xor
IC: Mod =   %cond = select i1 %cmp, i32 %or, i32 %xor
    New =   %cond = select i1 %cmp, i32 %or, i32 %xor
ADD:   %cond = select i1 %cmp, i32 %or, i32 %xor
ADD:   %or = or i32 -1, -1
IC: Visiting:   %or = or i32 -1, -1
IC: Replacing   %or = or i32 -1, -1
    with i32 -1
IC: Mod =   %or = or i32 -1, -1
    New =   %or = or i32 -1, -1
IC: ERASE   %or = or i32 -1, -1
IC: Visiting:   %cond = select i1 %cmp, i32 -1, i32 %xor
```

Reply:
I would say leave as todo in this patch. I think it better belongs more generically in `foldOpIntoSelect`.